### DISCUSSION TITLE
FastAPI resource leak when deploying on Vercel

### DISCUSSION THREAD

**Discussion by jackalnom** [fastapi/fastapi#9991] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
import os
import asyncio


app = FastAPI()

@app.get("/files_open/")
def files_open():
    fd_directory = "/proc/self/fd"
    fds = os.listdir(fd_directory)
    fd_info = {}
    for fd in fds:
        try:
            fd_info[fd] = os.readlink(os.path.join(fd_directory, fd))
        except FileNotFoundError:
            fd_info[fd] = "File descriptor closed before it could be read"
    return fd_info
```

### Description

When I deploy any FastAPI service to vercel and then make repeated calls to that service, it eventually gives me the error:
[ERROR] OSError: [Errno 24] Too many open files

This happens even in this minimal case where I initialize a completely default FastAPI instance with a static route that just returns a string. I tried instrumenting what is happening as you can see in my files_open endpoint above and for whatever reason sockets just continue to accumulate on every call until it reaches the system limits.

When I run this exact same code on Deta though I don't get this issue. It doesn't accumulate open sockets over time and everything works appropriately.

Example of calling files_open repeatedly:
{
  "0": "/dev/null",
  "1": "pipe:[5317]",
  "2": "pipe:[5317]",
  "3": "socket:[5330]",
  "4": "socket:[4477]",
  "5": "socket:[4478]",
  "6": "socket:[4479]",
  "7": "anon_inode:[eventpoll]",
  "8": "socket:[5498]",
  "9": "socket:[5499]",
  "10": "anon_inode:[eventpoll]",
  "11": "socket:[4644]",
  "12": "socket:[4645]",
  "13": "anon_inode:[eventpoll]",
  "14": "socket:[4654]",
  "15": "socket:[4655]",
  "16": "anon_inode:[eventpoll]",
  "17": "socket:[4662]",
  "18": "socket:[4663]",
  "19": "anon_inode:[eventpoll]",
  "20": "socket:[5524]",
  "21": "socket:[5525]",
  "22": "anon_inode:[eventpoll]",
  "23": "socket:[5481]",
  "24": "socket:[5482]",
  "25": "anon_inode:[eventpoll]",
  "26": "socket:[4676]",
  "27": "socket:[4677]",
  "28": "anon_inode:[eventpoll]",
  "29": "socket:[5542]",
  "30": "socket:[5543]",
  "31": "anon_inode:[eventpoll]",
  "32": "socket:[4686]",
  "33": "socket:[4687]",
  "34": "anon_inode:[eventpoll]",
  "35": "socket:[4688]",
  "36": "socket:[4689]",
  "37": "anon_inode:[eventpoll]",
  "38": "socket:[5570]",
  "39": "socket:[5571]",
  "40": "anon_inode:[eventpoll]",
  "41": "socket:[5574]",
  "42": "socket:[5575]",
  "43": "anon_inode:[eventpoll]",
  "44": "socket:[5584]",
  "45": "socket:[5585]",
  "46": "anon_inode:[eventpoll]",
  "47": "socket:[5598]",
  "48": "socket:[5599]",
  "49": "anon_inode:[eventpoll]",
  "50": "socket:[5622]",
  "51": "socket:[5623]",
  "52": "anon_inode:[eventpoll]",
  "53": "socket:[4734]",
  "54": "socket:[4735]",
  "55": "anon_inode:[eventpoll]",
  "56": "socket:[4736]",
  "57": "socket:[4737]",
  "58": "anon_inode:[eventpoll]",
  "59": "socket:[4742]",
  "60": "socket:[4743]",
  "61": "anon_inode:[eventpoll]",
  "62": "socket:[4748]",
  "63": "socket:[4749]",
  "64": "anon_inode:[eventpoll]",
  "65": "socket:[5666]",
  "66": "socket:[5667]",
  "67": "anon_inode:[eventpoll]",
  "68": "socket:[4754]",
  "69": "socket:[4755]",
  "70": "anon_inode:[eventpoll]",
  "71": "socket:[5684]",
  "72": "socket:[5685]",
  "73": "anon_inode:[eventpoll]",
  "74": "socket:[4764]",
  "75": "socket:[4765]",
  "76": "anon_inode:[eventpoll]",
  "77": "socket:[5701]",
  "78": "socket:[5702]",
  "79": "anon_inode:[eventpoll]",
  "80": "socket:[4775]",
  "81": "socket:[4776]",
  "82": "anon_inode:[eventpoll]",
  "83": "socket:[5719]",
  "84": "socket:[5720]",
  "85": "anon_inode:[eventpoll]",
  "86": "socket:[5724]",
  "87": "socket:[5725]",
  "88": "anon_inode:[eventpoll]",
  "89": "socket:[4794]",
  "90": "socket:[4795]",
  "91": "anon_inode:[eventpoll]",
  "92": "socket:[5742]",
  "93": "socket:[5743]",
  "94": "anon_inode:[eventpoll]",
  "95": "socket:[4804]",
  "96": "socket:[4805]",
  "97": "anon_inode:[eventpoll]",
  "98": "socket:[5760]",
  "99": "socket:[5761]",
  "100": "anon_inode:[eventpoll]",
  "101": "socket:[4813]",
  "102": "socket:[4814]",
  "103": "anon_inode:[eventpoll]",
  "104": "socket:[4818]",
  "105": "socket:[4819]",
  "106": "anon_inode:[eventpoll]",
  "107": "socket:[4827]",
  "108": "socket:[4828]",
  "109": "anon_inode:[eventpoll]",
  "110": "socket:[4837]",
  "111": "socket:[4838]",
  "112": "anon_inode:[eventpoll]",
  "113": "socket:[5791]",
  "114": "socket:[5792]",
  "115": "anon_inode:[eventpoll]",
  "116": "socket:[4853]",
  "117": "socket:[4854]",
  "118": "anon_inode:[eventpoll]",
  "119": "socket:[5805]",
  "120": "socket:[5806]",
  "121": "anon_inode:[eventpoll]",
  "122": "socket:[4863]",
  "123": "socket:[4864]",
  "124": "anon_inode:[eventpoll]",
  "125": "socket:[5824]",
  "126": "socket:[5825]",
  "127": "anon_inode:[eventpoll]",
  "128": "socket:[5830]",
  "129": "socket:[5831]",
  "130": "anon_inode:[eventpoll]",
  "131": "socket:[4891]",
  "132": "socket:[4892]",
  "133": "anon_inode:[eventpoll]",
  "134": "socket:[5843]",
  "135": "socket:[5844]",
  "136": "anon_inode:[eventpoll]",
  "137": "socket:[5850]",
  "138": "socket:[5851]",
  "139": "anon_inode:[eventpoll]",
  "140": "socket:[5862]",
  "141": "socket:[5863]",
  "142": "anon_inode:[eventpoll]",
  "143": "socket:[5869]",
  "144": "socket:[5870]",
  "145": "anon_inode:[eventpoll]",
  "146": "socket:[5876]",
  "147": "socket:[5877]",
  "148": "anon_inode:[eventpoll]",
  "149": "socket:[4939]",
  "150": "socket:[4940]",
  "151": "anon_inode:[eventpoll]",
  "152": "socket:[4949]",
  "153": "socket:[4950]",
  "154": "anon_inode:[eventpoll]",
  "155": "socket:[4953]",
  "156": "socket:[4954]",
  "157": "anon_inode:[eventpoll]",
  "158": "socket:[5911]",
  "159": "socket:[5912]",
  "160": "anon_inode:[eventpoll]",
  "161": "socket:[4971]",
  "162": "socket:[4972]",
  "163": "anon_inode:[eventpoll]",
  "164": "socket:[4985]",
  "165": "socket:[4986]",
  "166": "anon_inode:[eventpoll]",
  "167": "socket:[5919]",
  "168": "socket:[5920]",
  "169": "anon_inode:[eventpoll]",
  "170": "socket:[5929]",
  "171": "socket:[5930]",
  "172": "anon_inode:[eventpoll]",
  "173": "socket:[5939]",
  "174": "socket:[5940]",
  "175": "anon_inode:[eventpoll]",
  "176": "socket:[5949]",
  "177": "socket:[5950]",
  "178": "anon_inode:[eventpoll]",
  "179": "socket:[5013]",
  "180": "socket:[5014]",
  "181": "anon_inode:[eventpoll]",
  "182": "socket:[5967]",
  "183": "socket:[5968]",
  "184": "anon_inode:[eventpoll]",
  "185": "socket:[5983]",
  "186": "socket:[5984]",
  "187": "anon_inode:[eventpoll]",
  "188": "socket:[6177]",
  "189": "socket:[6178]",
  "190": "anon_inode:[eventpoll]",
  "191": "socket:[6034]",
  "192": "socket:[6035]",
  "193": "anon_inode:[eventpoll]",
  "194": "socket:[6187]",
  "195": "socket:[6188]",
  "196": "anon_inode:[eventpoll]",
  "197": "socket:[6052]",
  "198": "socket:[6053]",
  "199": "anon_inode:[eventpoll]",
  "200": "socket:[6066]",
  "201": "socket:[6067]",
  "202": "anon_inode:[eventpoll]",
  "203": "socket:[6076]",
  "204": "socket:[6077]",
  "205": "anon_inode:[eventpoll]",
  "206": "socket:[6197]",
  "207": "socket:[6198]",
  "208": "anon_inode:[eventpoll]",
  "209": "socket:[6093]",
  "210": "socket:[6094]",
  "211": "anon_inode:[eventpoll]",
  "212": "socket:[6103]",
  "213": "socket:[6104]",
  "214": "anon_inode:[eventpoll]",
  "215": "socket:[6117]",
  "216": "socket:[6118]",
  "217": "anon_inode:[eventpoll]",
  "218": "socket:[6123]",
  "219": "socket:[6124]",
  "220": "anon_inode:[eventpoll]",
  "221": "socket:[6227]",
  "222": "socket:[6228]",
  "223": "anon_inode:[eventpoll]",
  "224": "socket:[6134]",
  "225": "socket:[6135]",
  "226": "anon_inode:[eventpoll]",
  "227": "socket:[6249]",
  "228": "socket:[6250]",
  "229": "anon_inode:[eventpoll]",
  "230": "socket:[6140]",
  "231": "socket:[6141]",
  "232": "anon_inode:[eventpoll]",
  "233": "socket:[7171]",
  "234": "socket:[7172]",
  "235": "anon_inode:[eventpoll]",
  "236": "socket:[7177]",
  "237": "socket:[7178]",
  "238": "anon_inode:[eventpoll]",
  "239": "socket:[6280]",
  "240": "socket:[6281]",
  "241": "anon_inode:[eventpoll]",
  "242": "socket:[7195]",
  "243": "socket:[7196]",
  "244": "anon_inode:[eventpoll]",
  "245": "socket:[7202]",
  "246": "socket:[7203]",
  "247": "anon_inode:[eventpoll]",
  "248": "socket:[7210]",
  "249": "socket:[7211]",
  "250": "anon_inode:[eventpoll]",
  "251": "socket:[7218]",
  "252": "socket:[7219]",
  "253": "anon_inode:[eventpoll]",
  "254": "socket:[7228]",
  "255": "socket:[7229]",
  "256": "anon_inode:[eventpoll]",
  "257": "socket:[7243]",
  "258": "socket:[7244]",
  "259": "anon_inode:[eventpoll]",
  "260": "socket:[7250]",
  "261": "socket:[7251]",
  "262": "anon_inode:[eventpoll]",
  "263": "socket:[7257]",
  "264": "socket:[7258]",
  "265": "anon_inode:[eventpoll]",
  "266": "socket:[7263]",
  "267": "socket:[7264]",
  "268": "anon_inode:[eventpoll]",
  "269": "socket:[6350]",
  "270": "socket:[6351]",
  "271": "anon_inode:[eventpoll]",
  "272": "socket:[6360]",
  "273": "socket:[6361]",
  "274": "anon_inode:[eventpoll]",
  "275": "socket:[7291]",
  "276": "socket:[7292]",
  "277": "anon_inode:[eventpoll]",
  "278": "socket:[7308]",
  "279": "socket:[7309]",
  "280": "anon_inode:[eventpoll]",
  "281": "socket:[6363]",
  "282": "socket:[6364]",
  "283": "anon_inode:[eventpoll]",
  "284": "socket:[7334]",
  "285": "socket:[7335]",
  "286": "anon_inode:[eventpoll]",
  "287": "socket:[6373]",
  "288": "socket:[6374]",
  "289": "anon_inode:[eventpoll]",
  "290": "socket:[6382]",
  "291": "socket:[6383]",
  "292": "anon_inode:[eventpoll]",
  "293": "socket:[7346]",
  "294": "socket:[7347]",
  "295": "anon_inode:[eventpoll]",
  "296": "socket:[7352]",
  "297": "socket:[7353]",
  "298": "anon_inode:[eventpoll]",
  "299": "socket:[7364]",
  "300": "socket:[7365]",
  "301": "anon_inode:[eventpoll]",
  "302": "socket:[7371]",
  "303": "socket:[7372]",
  "304": "anon_inode:[eventpoll]",
  "305": "socket:[6420]",
  "306": "socket:[6421]",
  "307": "anon_inode:[eventpoll]",
  "308": "socket:[7378]",
  "309": "socket:[7379]",
  "310": "anon_inode:[eventpoll]",
  "311": "socket:[7390]",
  "312": "socket:[7391]",
  "313": "anon_inode:[eventpoll]",
  "314": "socket:[6443]",
  "315": "socket:[6444]",
  "316": "anon_inode:[eventpoll]",
  "317": "socket:[7401]",
  "318": "socket:[7402]",
  "319": "anon_inode:[eventpoll]",
  "320": "socket:[7408]",
  "321": "socket:[7409]",
  "322": "anon_inode:[eventpoll]",
  "323": "socket:[7420]",
  "324": "socket:[7421]",
  "325": "File descriptor closed before it could be read"
}

### Operating System

Linux

### Operating System Details

This is running on vercel with runs on lambda under the hood.

### FastAPI Version

0.46.0

### Python Version

3.9

**Comment by jgould22** [comment]
The first thing I would start with is updating the version of FastAPI as 0.46.0 is nearly 3.5 years old and seeing if that solves the issue.

**Reply by jackalnom** [reply]
Great suggestion. I upgraded FastAPI to 0.100.1. I still replicate the issue though unfortunately. 

For completeness, I dumped all packages installed on lambda with their versions:
{
  "message": [
    {
      "dist": "botocore",
      "version": "1.29.90",
      "size_in_mb": 74.582734
    },
    {
      "dist": "pip",
      "version": "22.0.4",
      "size_in_mb": 11.8048
    },
    {
      "dist": "setuptools",
      "version": "58.1.0",
      "size_in_mb": 3.239868
    },
    {
      "dist": "pydantic",
      "version": "2.1.1",
      "size_in_mb": 2.34009
    },
    {
      "dist": "boto3",
      "version": "1.26.90",
      "size_in_mb": 1.175309
    },
    {
      "dist": "urllib3",
      "version": "1.26.11",
      "size_in_mb": 0.683815
    },
    {
      "dist": "click",
      "version": "8.1.6",
      "size_in_mb": 0.644537
    },
    {
      "dist": "anyio",
      "version": "3.7.1",
      "size_in_mb": 0.5767
    },
    {
      "dist": "simplejson",
      "version": "3.17.2",
      "size_in_mb": 0.517367
    },
    {
      "dist": "s3transfer",
      "version": "0.6.0",
      "size_in_mb": 0.511988
    },
    {
      "dist": "fastapi",
      "version": "0.100.1",
      "size_in_mb": 0.478264
    },
    {
      "dist": "idna",
      "version": "3.4",
      "size_in_mb": 0.46314
    },
    {
      "dist": "starlette",
      "version": "0.27.0",
      "size_in_mb": 0.436496
    },
    {
      "dist": "uvicorn",
      "version": "0.23.2",
      "size_in_mb": 0.327033
    },
    {
      "dist": "h11",
      "version": "0.14.0",
      "size_in_mb": 0.309589
    },
    {
      "dist": "jmespath",
      "version": "1.0.1",
      "size_in_mb": 0.10778
    },
    {
      "dist": "exceptiongroup",
      "version": "1.1.2",
      "size_in_mb": 0.057606
    },
    {
      "dist": "awslambdaric",
      "version": "2.0.4",
      "size_in_mb": 0.047096
    },
    {
      "dist": "sniffio",
      "version": "1.3.0",
      "size_in_mb": 0.010865
    }
  ]
}
